codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ServiceGenerator.java

```diff
@@ -177,6 +178,71 @@ private void generateApplicationProtocolTypes() {
         }).write("");
     }
 
+    private void generateClientInvokeOperation() {
+        writer.addUseImports(SmithyGoDependency.CONTEXT);
+        writer.addUseImports(SmithyGoDependency.SMITHY);
+
+        writer.openBlock("func (c *Client) invokeOperation("
+                + "ctx context.Context, "
+                + "opID string, "
+                + "params interface{}, "
+                + "optFns []func(*Options), "
+                + "stackFns ...func(*middleware.Stack, Options) error"
+                + ") "
+                + "(result interface{}, metadata middleware.Metadata, err error) {", "}", () -> {
+            writer.addUseImports(SmithyGoDependency.SMITHY_MIDDLEWARE);
+            writer.addUseImports(SmithyGoDependency.SMITHY_HTTP_TRANSPORT);
+
+            generateConstructStack();
+            writer.write("options := c.options.Copy()");
+            writer.write("for _, fn := range optFns { fn(&options) }");
+            writer.write("");
+
+            writer.openBlock("for _, fn := range stackFns {", "}", () -> {
+                writer.write("if err := fn(stack, options); err != nil { return nil, metadata, err }");
+            });
+            writer.write("");
+
+            writer.openBlock("for _, fn := range options.APIOptions {", "}", () -> {
+                writer.write("if err := fn(stack); err != nil { return nil, metadata, err }");
+            });
+            writer.write("");
+
+            generateConstructStackHandler();
+            writer.write("result, metadata, err = handler.Handle(ctx, params)");
+            writer.openBlock("if err != nil {", "}", () -> {
+                writer.openBlock("err = &smithy.OperationError{", "}", () -> {
+                    writer.write("ServiceID: ServiceID,");
+                    writer.write("OperationName: opID,");
+                    writer.write("Err: err,");
+                });
```

Comment:
Consider renaming `OperationError` with `OperationInvokeError` to reduce confusion with `APIError`

Reply:
We probably should do that in a separate PR since that's a pretty wide reaching change.

Reply:
Created https://github.com/awslabs/smithy-go/issues/210 for this.